doc/man/gnome-session-quit.1
.\"
.\" gnome-session-quit manual page.
.\" (C) 2000 Miguel de Icaza (miguel@helixcode.com)
.\" (C) 2009-2010 Vincent Untz (vuntz@gnome.org)
.\"
.TH GNOME-SESSION-QUIT 1 "GNOME"
.SH NAME
gnome-session-quit \- End the current GNOME session
.SH SYNOPSIS
.B gnome-session-quit [\-\-logout|\-\-power-off] [\-\-no-prompt]
.SH DESCRIPTION
The \fIgnome-session-quit\fP program can be used to end a GNOME session.
.SH OPTIONS
The following options are supported:
.TP
.I "--logout"
Prompt the user to confirm logout. This is the default behavior.
.TP
.I "--power-off"
Prompt ther user to confirm system power off.
.TP
.I "--no-prompt"
End the session without user interaction. This only works with \fI--logout\fP.
.SH BUGS
If you find bugs in the \fIgnome-session-quit\fP program, please report
these on https://bugzilla.gnome.org.
.SH SEE ALSO
.BR gnome-session(1)


NEWS
===============
Version 2.91.94
===============

Session Manager: Bump the phase timeout to 30 seconds.

Translations: Swedish, Ukrainian, Punjabi, Norwegian bokmål,
  Bazilian Portuguese, Dutch, Hebrew, German, Japanese, Arabic,
  Russian, Vietnamese, Uighur, French, Latvian, Polish

===============
Version 2.91.93
===============

Note this is a brown bag release.  All users who need working log out and shut down
functionality should use this release instead of 2.91.92.

Session Manager:
- Make work with latest version of the shell.

Translations:
- Updated bg: Krasimir Chonov
- Updated es: Jorge González
- Updated it: Luca Ferretti
- Updated ro: Daniel Șerbănescu
- Updated sl: Matej Urbančič

===============
Version 2.91.92
===============

Note to packagers: if you are using your own .session files, please update them
as the format has changed.

Session Manager:
- Better fatal error handling in general, including specific handling when
  required components fail (Colin Walters, William Jon McCann, Vincent)
- Improve wording of fallback dialog (Matthias Clasen)
- Avoid double-checking between gdm and gnome-session in
  gnome-session-check-accelerated-helper (Matthias Clasen)
- Switch to using seconds for idle time (William Jon McCann)
- Honor disable-log-out and disable-user-switching lockdown settings (Ray
  Strode, Vincent)
- Change format of .session files in a incompatible way; this is now documented
  in the gnome-session man page (Vincent)
- Update gnome-session-quit man page (Vincent)
- Enforce the use of required components in the session (Vincent)
- Deal with conflicts between Provides in different applications by only having
  one provider running; this was only working well for Provides defined in the
  session definition (Vincent)
- Use app-id when saving session instead of startup-id for .desktop (Vincent)
- Explicitly require gnome-shell/gnome-panel in .session files, instead of
  using providers (Vincent)
- Clean up saved session on logout if session is not getting saved (Vincent)
- Use "computer-fail" icon from theme instead of our own sad-computer (Vincent)
- Fix some gnome-shell/gnome-session interaction issue on logout, when a third
  application cancels logout (Ray Strode)
- Code cleanups (Ray Strode, Vincent)

Translations:
- Updated ar: Khaled Hosny
- Updated as: Nilamdyuti Goswami
- Updated en_GB: Bruce Cowan
- Updated es: Jorge González
- Updated et: Ivar Smolin
- Updated fr: Bruno Brouard
- Updated fy: Sense Hofstede
- Updated gl: Fran Diéguez
- Updated he: Yaron Shahrabani
- Updated hu: Gabor Kelemen
- Updated id: Andika Triwidada
- Updated ko: Changwoo Ryu
- Updated lv: Rudolfs
- Updated nb: Kjartan Maraas
- Updated nl: Wouter Bolsterlee
- Updated pl: Piotr Drąg
- Updated pt: Duarte Loreto
- Updated pt_BR: Flamarion Jorge
- Updated ro: Lucian Adrian Grijincu
- Updated ru: Yuri Myaseodov
- Updated sl: Andrej Žnidaršič
- Updated sr: Милош Поповић
- Updated sr@latin: Miloš Popović
- Updated sv: Daniel Nylander
- Updated uk: Korostil Daniel

=================
Version 2.91.91.3
=================

Fix another problem with launching of gnome-session-check-accelerated-helper

=================
Version 2.91.91.2
=================

More fixes and diagnostics for gnome-session-check-accelerated

=================
Version 2.91.91.1
=================

Fix gnome-session-check-accelerated to return 0 on success, as expected

===============
Version 2.91.91
===============

Session Manager:
- Better error message on not-found session (Colin Walters)
- Rename gnome-session-save to gnome-session-quit (William Jon McCann)
- Fix cancel button in logout dialog (Ray Strode)
- Increase is-accelerated runnable helper timeout (Colin Walters)
- Fix crash when if-session autostart condition is used (Matthias Clasen)
- Add gnome-session-check-accelerated to set a X property about acceleration
  (Colin Walters)
- Rename gnome-session-is-accelerated to gnome-session-check-accelerated-helper
  (Colin Walters)
- Fix typo (Vincent)

Translations:
- Updated ar: Khaled Hosny
- Updated bg: Alexander Shopov
- Updated ca: David Planella
- Updated es: Daniel Mustieles
- Updated et: Ivar Smolin
- Updated fr: Alain Lojewski
- Updated he: Yaron Shahrabani
- Updated ko: Changwoo Ryu
- Updated lv: Rūdolfs Mazurs
- Updated nb: Kjartan Maraas
- Updated pa: A S Alam
- Updated sk: Pavol Šimo
- Updated sl: Matej Urbančič
- Updated sv: Daniel Nylander
- Updated uk: Korostil Daniel

=================
Version 2.91.90.1
=================

Session Manager:
- Fix a crash happening in gdm (Colin Walters)

Translations:
- Updated bn_IN: Runa Bhattacharjee

===============
Version 2.91.90
===============

Session Manager:
- Update for GSettings schemas changes (Vincent)
- Read default session from GSettings (Alexander Larsson)
- Make the current session name available via dbus (Alexander Larsson)
- Support autostart conditions that depend on the current session (Alexander
  Larsson)
- Make notification-daemon a required part of the fallback session (Matthias
  Clasen)
- Show a dialog after login the first time we fall back to gnome-fallback
  (Alexander Larsson)
- Various fixes to shell logout/shutdown dialog (Ray Strode)
- Code cleanups (Vincent)

Misc:
- gsettings-desktop-schemas 0.1.7 is required

Translations:
- Updated ar: Khaled Hosny
- Updated bg: Alexander Shopov
- Updated ca: David Planella
- Updated es: Jorge González
- Updated gu: Sweta Kothari
- Updated he: Yaron Shahrabani
- Updated nb: Kjartan Maraas
- Updated nl: Wouter Bolsterlee
- Updated sl: Matej Urbančič

==============
Version 2.91.6
==============

Note to packagers: with the move to GSettings, gnome-session now depends on
gsettings-desktop-schemas.

Session Manager:
- Fix build with latest GTK+ 3 (Matthias Clasen, Vincent)
- Migrate to GSettings (Ray Strode)
- Add support for shell logout/shutdown dialog (Ray Strode)
- Use fallback session if required components are missing (Matthias Clasen)

Defined Sessions:

tools/gnome-session-quit.c
/* -*- Mode: C; tab-width: 8; indent-tabs-mode: nil; c-basic-offset: 8 -*-
 * save-session.c - Small program to talk to session manager.

   Copyright (C) 1998 Tom Tromey
   Copyright (C) 2008 Red Hat, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
   02111-1307, USA.
*/

#include <config.h>

#include <unistd.h>
#include <stdlib.h>
#include <stdio.h>
#include <string.h>

#include <glib/gi18n.h>
#include <gtk/gtk.h>

#include <dbus/dbus-glib.h>
#include <dbus/dbus-glib-lowlevel.h>

#define GSM_SERVICE_DBUS   "org.gnome.SessionManager"
#define GSM_PATH_DBUS      "/org/gnome/SessionManager"
#define GSM_INTERFACE_DBUS "org.gnome.SessionManager"

enum {
        GSM_LOGOUT_MODE_NORMAL = 0,
        GSM_LOGOUT_MODE_NO_CONFIRMATION,
        GSM_LOGOUT_MODE_FORCE
};

static gboolean logout = FALSE;
static gboolean power_off = FALSE;
static gboolean no_prompt = FALSE;
static gboolean force = FALSE;

static GOptionEntry options[] = {
        {"logout", '\0', 0, G_OPTION_ARG_NONE, &logout, N_("Log out"), NULL},
        {"power-off", '\0', 0, G_OPTION_ARG_NONE, &power_off, N_("Power off"), NULL},
        {"force", '\0', 0, G_OPTION_ARG_NONE, &force, N_("Ignoring any existing inhibitors"), NULL},
        {"no-prompt", '\0', 0, G_OPTION_ARG_NONE, &no_prompt, N_("Don't prompt for user confirmation"), NULL},
        {NULL}
};

static void
display_error (const char *message)
{
        g_printerr ("%s\n", message);
}

static DBusGConnection *
get_session_bus (void)
{
        DBusGConnection *bus;
        GError *error = NULL;

        bus = dbus_g_bus_get (DBUS_BUS_SESSION, &error);

        if (bus == NULL) {
                g_warning ("Couldn't connect to session bus: %s", error->message);
                g_error_free (error);
        }

        return bus;
}

static DBusGProxy *
get_sm_proxy (void)
{
        DBusGConnection *connection;
        DBusGProxy      *sm_proxy;

        connection = get_session_bus ();
        if (connection == NULL) {
                display_error (_("Could not connect to the session manager"));
                return NULL;
        }

        sm_proxy = dbus_g_proxy_new_for_name (connection,
                                               GSM_SERVICE_DBUS,
                                               GSM_PATH_DBUS,
                                               GSM_INTERFACE_DBUS);

        if (sm_proxy == NULL) {
                display_error (_("Could not connect to the session manager"));
                return NULL;
        }

        return sm_proxy;
}

static void
do_logout (unsigned int mode)
{
        DBusGProxy *sm_proxy;
        GError     *error;
        gboolean    res;

        sm_proxy = get_sm_proxy ();
        if (sm_proxy == NULL) {
                return;
        }

        error = NULL;
        res = dbus_g_proxy_call (sm_proxy,
                                 "Logout",
                                 &error,
                                 G_TYPE_UINT, mode,
                                 G_TYPE_INVALID,
                                 G_TYPE_INVALID);

        if (!res) {
                if (error != NULL) {
                        g_warning ("Failed to call logout: %s",
                                   error->message);
                        g_error_free (error);
                } else {
                        g_warning ("Failed to call logout");
                }
        }

        if (sm_proxy != NULL) {
                g_object_unref (sm_proxy);
        }
}

static void
do_power_off (void)
{
        DBusGProxy *sm_proxy;
        GError     *error;
        gboolean    res;

        sm_proxy = get_sm_proxy ();
        if (sm_proxy == NULL) {
                return;
        }

        error = NULL;
        res = dbus_g_proxy_call (sm_proxy,
                                 "Shutdown",
                                 &error,
                                 G_TYPE_INVALID,
                                 G_TYPE_INVALID);

        if (!res) {
                if (error != NULL) {
                        g_warning ("Failed to call shutdown: %s",
                                   error->message);
                        g_error_free (error);
                } else {
                        g_warning ("Failed to call shutdown");
                }
        }

        if (sm_proxy != NULL) {
                g_object_unref (sm_proxy);
        }
}

int
main (int argc, char *argv[])
{
        GError *error;
        int     conflicting_options;

        /* Initialize the i18n stuff */
        bindtextdomain (GETTEXT_PACKAGE, LOCALE_DIR);
        bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
        textdomain (GETTEXT_PACKAGE);

        error = NULL;
        if (! gtk_init_with_args (&argc, &argv, NULL, options, NULL, &error)) {
                g_warning ("Unable to start: %s", error->message);
                g_error_free (error);
                exit (1);
        }

        conflicting_options = 0;
        if (logout)
                conflicting_options++;
        if (power_off)
                conflicting_options++;
        if (conflicting_options > 1)
                display_error (_("Program called with conflicting options"));

        if (power_off) {
                do_power_off ();
        } else {
                /* default to logout */

                if (force)
                        do_logout (GSM_LOGOUT_MODE_FORCE);
                else if (no_prompt)
                        do_logout (GSM_LOGOUT_MODE_NO_CONFIRMATION);
                else
                        do_logout (GSM_LOGOUT_MODE_NORMAL);
        }

        return 0;
}
